1.fee_bumping/fee_bumping.md

```diff
@@ -153,25 +153,208 @@ Bitcoin blockchain, such as exchanges or custodians:
+section describes some of those implications.
+
+[five conditions]: https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#implementation-details
+
+### Implications of the conditions for BIP 125 RBF
+
+> One or more transactions currently in the mempool (original transactions)
+> will be replaced by a new transaction (replacement transaction) that spends
+> one or more of the same inputs if,
+
+- implication 1: RBF can be used to replace multiple transactions at the same
+  time. That could either be because the original transaction has descendants
+  in the mempool that would be invalidated by the replacement transaction and
+  so need to be removed, or because the replacement transaction is spending
+  inputs in more than one original transaction.
+- implication 2: Additional inputs can be added to the replacement transaction.
+  This is useful if a wallet needs to bump the fee on a transaction but there
+  is no change output to subtract the additional fee from.
+- implication 3: The replacement transaction does not need to include all of
+  the inputs from the original transaction. This can be extremely dangerous
+  since wallets can easily overspend this way. For example, if the original
+  transaction uses inputs A and B, a replacement transaction is sent with
+  inputs B and C, and then that replacement transaction is replaced with a
+  new replacement transaction using inputs C and D, the first transaction
+  could be rebroadcast by a third party at any point in the future, and both
+  would be valid! Wallets are advised to include all inputs from the original
+  transaction in the replacement transaction unless they have very reliable
+  tracking to make sure that they don't double-spend themselves.
+- implication 4: the `nSequence` of the inputs in the replacement transaction
+  are not considered, so a replaceable transaction could be replaced by a
+  non-replaceable transaction. If a wallet wants to be able to replace the
+  replacement transaction in future, it must signal opt-in RBF on one of the
+  inputs of the replacement transaction.
+
+> 1. The original transactions signal replaceability explicitly or through
+> inheritance as described in the above Summary section.
+
+- implication 1: It is not sufficient to look at an individual transaction
+  to determine whether it can be replaced under this policy. A transaction
+  may be replaceable because one of its unconfirmed ancestors signals
+  opt-in RBF.
+- implication 2: A transaction's replaceability may change! If it is does
+  not signal opt-in replaceability, but one of its ancestors does, then it
+  is replaceable. If the ancestor then gets confirmed, the descendant
+  transaction becomes non-replaceable.
+
+> 2. The replacement transaction does not contain any new unconfirmed inputs
+> that did not previously appear in the mempool. (Unconfirmed inputs are
+> inputs spending outputs from currently unconfirmed transactions.)
```

Comment:
@dgpv says "This is really confusing description in a BIP125, especially this "did not previously appear in the mempool" part.

In the code, the logic is: Do not allow new unconfirmed inputs.
But to avoid UTXO set lookup, just check if the input is in a mempool. If it is - this means that it is unconfirmed - reject tx. If the input is not in mempool - either it is confirmed, or non-existent, but that just means that tx that depends on it will be invalid."

Reply:
Here the confusing part was "that did not previously appear in the mempool." Maybe its double-negation that is hard to understand: "does not contain .. that did not appear..."

Let's say we have an unconfirmed tx A that is already in the mempool.

tx B did not use outputs of tx A.

tx B' wants to use outputs of tx A as its inputs.

We can say that such inputs (the outputs of A) does indeed "previously appear in the mempool".

tx B' contain new unconfirmed inputs, but as this inputs "previously appear in the mempool" we could think that tx B' are fine.

But the logic in the code does not allow B' to use such new inputs - because they are unconfirmed.

Reply:
I do not think that language around input/output is much of a problem. At least in two cases where I was confused (here and multi-tx-replacement-at-same-level), it was another aspects of language that was the cause for my confusion, I think.

Reply:
The BIP author has suggested a change to the BIP125 text here: https://github.com/bitcoin/bips/pull/745.